include-what-you-use CI job and code fixes

N-Coder opened the pull request:
This PR has two parts. The [first part](https://github.com/ogdf/ogdf/compare/nfink-iwyu) adds iwyu to our clang docker images (also streamlining some of the image build process) and add scripts together with a CI job that automatically run iwyu against the ogdf codebase. The [second part](https://github.com/ogdf/ogdf/compare/nfink-iwyu...nfink-iwyu-fixes) then contains the changes suggested by iwyu plus some manual fixes that ensure iwyu does the right thing. This process needed a few iterations (which can be squashed before merging), mostly because I also needed some time to figure out how things (should) work and which parts of the OGDF need special treatment. Hopefully now everything should be stable in the sense that running iwyu and applying its fixes once (which the CI provides as patch file artifact) should be enough to make it happy for any future PRs.

**Open ToDos:**
- [x] let fix_includes.sh run iwyu in docker container if command is not available
- [x] ensure that docker images are deployed
- [x] switch all usages of ncoder/ to ogdf/ images (depends on point above)
- [ ] remove empty files left over from registered arrays PR or make them proxies
- [ ] announce "breaking change" that the transitively imported headers might have changed

milsen commented on util/docker/clang/Dockerfile:
```diff
@@ -13,9 +15,16 @@ RUN apt-get update \
+    clang-format-${llvmver} clang-tools-${llvmver} llvm-${llvmver}-dev lld-${llvmver} lldb-${llvmver}  \
+    llvm-${llvmver}-tools libomp-${llvmver}-dev libc++-${llvmver}-dev libc++abi-${llvmver}-dev  \
+    libclang-common-${llvmver}-dev libclang-${llvmver}-dev libclang-cpp${llvmver}-dev libunwind-${llvmver}-dev \
+ && apt-get install clang-format-15 \
```
This needs to be `clang-format-$llvmver`, otherwise apt cannot find clang-15 when building e.g. the image clang:18.

N-Coder replied:
`clang-format-15` is in the [debian bookworm package sources](https://pkgs.org/search/?q=clang-format-15), so this should build fine - maybe your cached version of `debian:latest` is outdated? The reason behind this is that I want a container that contains both a recent iwyu (using clang-18) and our clang-format-15, so that the CI can always apply both tools directly after each other.

milsen replied:
You were right, my `debian:latest` was outdated. I updated all the docker images now. Thank you!

N-Coder commented:
Ok, this should be good now from the technical side. The only thing that still needs discussing is how we want to handle this w.r.t. users. Should we for now only process our own source files and leaves headers as they are or otherwise how should we announce this "breaking" change? Note that we now only make it explicit that people shouldn't rely on transitive includes, previously these could also change without warning, potentially breaking users' code. By making this explicit we could also tell users to run iwyu on their own code before upgrading the OGDF, then missing transitive includes shouldn't be a problem for them no more. In that case, we could also simply drop all the old headers files that have now become empty (like NodeArray.h and the like).

N-Coder commented:
The agreement at GD was that we want to merge this PR once the discussion points from above are settled (:eyes: @milsen @wirkom). The changes to the imports look nice for now and if we ever encounter any major issues with the tool, we can simply disable the CI step again or make it optional. Once the Sync Plan PRs are merged, I can rebase this onto master and re-run iwyu a final time for any newly added code.

N-Coder commented:
Ok, this should be good to get merged by now. Let's see how iwyu fares when used in practice.
I'll create a separate PR for deleting the now-unused headers and adding a respective statement to the changelog.

One more idea: what would you think of an include-all header, say `<ogdf/basic/_all.h>`, that makes ensuring sufficient includes while developing easy (at the cost of slightly increased compilation times) and that can later (i.e., at latest for merging) easily be resolved to the actually needed files by iwyu?

milsen commented:
_Several weeks later..._ 
I read through the changes and tested this out myself (the docker images have been adapted). Thank you for all your work!

If you rebase, I'll merge. Some notes:
* In [fix_includes.py, line 199](https://github.com/ogdf/ogdf/blob/95b0917b503f88631dbdecded6628e7791095f55/util/fix_includes.sh#L199), we let iwyu ignore `cluster/internal/`. Is that (still) needed?
* As you said, these changes have to be noted in the porting guide, but if you want to do that in a separate PR, that's fine.
* As for the _all-header, I personally do not see a big need for it. I'd just add header files I currently need and fix them at the end with `fix_includes.py`. So the advantage of `_all.h` seems quite minor, but if you want it, we can add it.

N-Coder commented:
> In [fix_includes.py, line 199](https://github.com/ogdf/ogdf/blob/95b0917b503f88631dbdecded6628e7791095f55/util/fix_includes.sh#L199), we let iwyu ignore cluster/internal/. Is that (still) needed?

I recall that the `cluster/interal` code was pretty messy with its use of abacus and coin, so I at one point decided to skip it. But as the "fixes" commits actually did clean up that code and everything seemed to work fine anyways, I just rebased once again and dropped that filter. Once CI finishes, this should be good to be merged now.
